tools/docker/Dockerfile

```diff
@@ -0,0 +1,231 @@
+# syntax=docker/dockerfile:experimental
+# Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+# 	http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+#########################################
+#
+# BUILD IMAGES
+#
+#########################################
+
+
+# Common tools needed for the build stages ahead. The final test images do not inherit directly from here, so this bloat
+# is dropped in those final end-use images.
+FROM debian:stretch as build-base
+ENV PATH /bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin:/usr/lib/go/bin
+RUN --mount=type=cache,id=aptcache,target=/var/cache/apt \
+	--mount=type=cache,id=aptlib,target=/var/lib/apt,sharing=locked \
+	mkdir -p /etc/apt/sources.list.d \
+	&& echo "deb http://ftp.debian.org/debian stretch-backports main" > /etc/apt/sources.list.d/stretch-backports.list \
+	&& apt update \
+	&& apt --target-release stretch-backports install -y golang-go \
+	&& apt install -y \
+		git \
+		wget \
+		build-essential \
+		libseccomp-dev \
+		btrfs-tools \
+		libdevmapper-dev \
+		musl-tools \
+		util-linux \
+		ca-certificates
+
+# Run as non-root for the rest of the build stages now that we've gotten apt installs out of the way.
+RUN useradd --create-home builder \
+	&& mkdir /output \
+	&& chown builder /output \
+	&& mkdir -p /home/builder/go
+
+# Buildkit caches don't support anything like a "--chown" flag yet, so ensure builder will have access to them
+RUN --mount=id=gomod,target=/home/builder/go/pkg/mod,type=cache \
+	chown -R builder /home/builder/go
+
+USER builder
+WORKDIR /home/builder
+SHELL ["/bin/bash", "-c"]
+
+
+
+
+# Build firecracker itself
+FROM build-base as firecracker-build
+ENV RUSTUP_HOME=/home/builder/rustup \
+    CARGO_HOME=/home/builder/cargo \
+    PATH=/home/builder/cargo/bin:$PATH \
+	RUST_VERSION=1.32.0
+
+# TODO using ADD <url> should work instead of having to wget, but buildkit has a bug where --chown is ignored for URLs
+# https://github.com/moby/buildkit/issues/813
+RUN wget "https://static.rust-lang.org/rustup/archive/1.16.0/x86_64-unknown-linux-gnu/rustup-init" \
```

Comment:
I'd usually prefer curl over wget, with options such as `--max-time`, `--retry`, and `--silent` (maybe among others, depending on circumstances). You don't need to switch to curl, but it'd be nice make sure equivalent options are applied here (or not, if wget's defaults are reasonable).